amqp/src/main/scala/akka/stream/alpakka/amqp/scaladsl/AmqpFlowWithContext.scala

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2016-2019 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.amqp.scaladsl
+
+import akka.Done
+import akka.stream.alpakka.amqp._
+import akka.stream.scaladsl.{Flow, FlowWithContext}
+
+import scala.concurrent.Future
+import scala.concurrent.duration._
+
+object AmqpFlowWithContext {
+
+  /**
+   * Creates a contextual variant of corresponding [[AmqpFlow]].
+   *
+   * @see [[AmqpFlow.apply]]
+   */
+  def apply[T](
+      settings: AmqpWriteSettings
+  ): FlowWithContext[WriteMessage, T, WriteResult, T, Future[Done]] =
+    FlowWithContext.fromTuples(
+      Flow.fromGraph(new impl.AmqpSimpleFlowStage[T](settings))
+    )
+
+  /**
+   * Creates a contextual variant of corresponding [[AmqpFlow]].
+   *
+   * @see [[AmqpFlow.withConfirm]]
+   */
+  def withConfirm[T](
+      settings: AmqpWriteSettings,
+      confirmationTimeout: FiniteDuration
+  ): FlowWithContext[WriteMessage, T, WriteResult, T, Future[Done]] =
+    FlowWithContext.fromTuples(
+      Flow.fromGraph(new impl.AmqpBlockingFlowStage[T](settings, confirmationTimeout))
+    )
+
+  /**
+   * Creates a contextual variant of corresponding [[AmqpFlow]].
+   *
+   * @see [[AmqpFlow.withAsyncConfirm]]
+   */
+  def withAsyncConfirm[T](
+      settings: AmqpWriteSettings,
+      bufferSize: Int,
+      confirmationTimeout: FiniteDuration
+  ): FlowWithContext[WriteMessage, T, WriteResult, T, Future[Done]] =
+    FlowWithContext.fromTuples(
+      Flow.fromGraph(new impl.AmqpAsyncFlowStage(settings, bufferSize, confirmationTimeout))
+    )
+
+  /**
+   * Creates a contextual variant of corresponding [[AmqpFlow]].
+   *
+   * @see [[AmqpFlow.withAsyncUnorderedConfirm]]
+   */
+  def withAsyncUnorderedConfirm[T](
+      settings: AmqpWriteSettings,
+      bufferSize: Int,
+      confirmationTimeout: FiniteDuration
+  ): FlowWithContext[WriteMessage, T, WriteResult, T, Future[Done]] =
+    FlowWithContext.fromTuples(
+      Flow.fromGraph(new impl.AmqpAsyncUnorderedFlowStage(settings, bufferSize, confirmationTimeout))
+    )
```

Comment:
Hm, `FlowWithContext` is supposed to always keep message order. So I guess the unordered flow stage should not support context passing.
A motivation for this criteria are Kafka offsets which should not be reordered.

Reply:
Well, I never used Kafka connector, but here are my thoughts on that: it seems strange to enforce such ordering constraints since `FlowWithContext` can be used there to carry arbitrary objects, not only offsets (if I understand correctly). Offsets are a fundamental concept in Kafka so I would expect users to be familiar with it and having "unordered" somewhere in a name would be enough warning sign to check if that can be used with offsets (warning could be also included in docs ofc).

Nevertheless, regardless of Kafka, I don't see the reason to enforce such convention across all alpakka connectors. If I understand correctly `FlowWithContext` in general is not meant to preserve order (nothing about that in akka streams docs anyway) so unordered variants should not be suprising for users. Unordered flows are a quick way to boost performance and it would be sad to not have them with context 😉. 

Ofc I do not insist on that and I can remove it if you want so.

Reply:
You're right, the Akka documentation around `FlowWithContext` does hardly exist. It's in our plans to improve on it. Even for a normal Akka Stream keeping order is expected unless you go unordered explicitly (as you do here).

Taking a step back: In which scenarios does AMQP reply out of order to the published messages? Are you sure it provides significantly better performance so it is worth to have the unordered variants?
If it does, you may expose the `withAsyncUnorderedConfirm` as a tuple flow (which can easily be wrapped by the user).

Similarly, is the distinction between `withConfirm` and `withAsyncConfirm` really important? 
Wouldn't a buffer size of 1 effectively give `withConfirm` behaviour? The buffer size could become a value in the `AmqpWriteSettings`.
The `confirmationTimeout` could be moved to the settings, as well.

Reply:
Looking at the RabbitMQ docs right now to answer your questions and I just realized that confirmation mechanism I used is not a part of AMQP protocol but a RabbitMQ-specific extension to it. I guess that disqualifies all variants except for basic one that don't use it. Am I right?


Anyway, my thoughts on your concerns:

When exactly messages are confirmed by broker depends on few things. It's described shortly in [docs](https://www.rabbitmq.com/confirms.html#when-publishes-are-confirmed), but most notably for persistent messages acks are sent only after message is persisted to disk (which is done in batches, so the latency here may vary) and for mirrored queues messages are ack'ed if all mirrors accept the message, so the network unreliability is involved here. I can imagine scenario where I would expect unordered flow would be significantly more performant, but I admit I haven't done any tests to prove that.
I am not pleased by inconsistency of having unordered variant as a tuple flow, but if there is no other way then I'm ok with that 😉.

Indeed, `withAsyncConfirm` with buffer size of 1 should give something similar as `withConfirm`. I haven't thought about that. It can be removed.

I didn't want to add buffer size and timeout to `AmqpWriteSettings` because now it contains settings common to all flows and sinks, while those two are more specific. I saw the same approach used in `AmqpRpcFlow` as well.

Reply:
Sorry for my late reply.

> I guess that disqualifies all variants except for basic one that don't use it. Am I right?

The RabbitMQ specific thing is a problem. But I wonder if anything but RabbitMQ is used somewhere... I assume it is the most common AMQP broker so we might mark those flows as specific to RabbitMQ. But we need to make sure the existing ones are not broken.

> I am not pleased by inconsistency of having unordered variant as a tuple flow, but if there is no other way then I'm ok with that 😉.

I understand that, but for now, it's the best solution I see.

> I didn't want to add buffer size and timeout to AmqpWriteSettings because now it contains settings common to all flows and sinks

That is always a balance, but settings are the solution to add more later without breaking the API. It is better to have settings that are not used in certain factories.

Reply:
I removed the `withConfirm` variants and adjusted the namings. Unordered flow with context is removed and replaced by flow of pairs, as discussed. I moved the `bufferSize` and `confirmationTimeout` to `WriteSettings` but I didn't apply this to rpc flow - I don't know what are the plans to release this so I didn't want to introduce breaking change there.

Please check out the changes.